### ISSUE TITLE
Requirements for PodDisruptionBudget, /eviction subresource, and PreferAvoidPods to graduate to Beta and then v1

### ISSUE THREAD

**Issue by davidopp** [kubernetes/kubernetes#25321]
We'd like to graduate PodDisruptionBudget, /eviction subresource, and PreferAvoidPods to Beta in 1.5

PodDisruptionBudget to Beta
- [ ] Make sure we can add disruption rate later in a backward-compatible way (don't need to implement it now)
- [ ] Add PodDisruptionBudget as a spreading criteria in the scheduler
- [x] Move to beta API group

PodDisruptionBudget to v1
- [ ] Move to v1 API group

PreferAvoidPods to Beta
- [ ] Finish [AvoidPods](https://github.com/kubernetes/kubernetes/blob/3fd14d97fb13ba2849e0c908aaff18efcece70c1/pkg/api/types.go#L2146) feature: add a "avoid same node" bool to eviction request, which causes API server to add the pod's current node to this annotation on the Node. PDB controller should age old out entries in the list according to some policy. (The scheduler part of this feature is implemented, i.e. see `NodePreferAvoidPodsPriority` priority function.)
- [ ] Change annotation key from alpha to beta

PreferAvoidPods to v1
- [ ] Move AvoidPods from annotation to field of NodeStatus
  /evict subresource to Beta

/eviction subresource to Beta
- [ ] Have `kubectl drain` use /eviction subresource, including "avoid rescheduling on current node"  behavior (see PreferAvoid Pods description above) (not strictly required to move to Beta, but having a Proof of Concept use of the feature that people can look at is important, so I think it's reasonable to make this part of moving it to Beta)
- [ ] Move to beta API group

/evict subresource to v1
- [ ] Move to v1 API group

(For all Betas, make sure we have a rollback story or are sure we won't need to rollback earlier than when feature was first introduced.)

cc/ @mml @bgrant0607 @jiangyaoguo 

ref/ #22217 #12611 #24697 #20699 #18853

**Comment by ConnorDoyle** [comment]
xref PR #31033

**Comment by davidopp** [comment]
Need to look into this (from https://github.com/kubernetes/kubernetes.github.io/pull/1244)

I would rewrite this as

Currently a PodDisruptionBudget has two components: a selector to specify a set of pods, and a description of the minimum number of pods from that set that must be available (i.e. an eviction will not be allowed if it will cause the number of available pods to fall below this threshold).
       @mml
mml 2 hours ago
The "i.e." is not how it's implemented. This may be a simple off-by-one error, or it may be a place where we didn't think carefully enough about the spec.

As it stands, if you say "100%", we will allow evictions if all the pods are ready. If we want "100%" to mean "never permit a 'voluntary' eviction", I will clarify the docs, and changing the code and tests will be easy.

Or maybe the right thing is to open an issue to discuss and track this, and for now document the way it works?

**Comment by balajismaniam** [comment]
@davidopp Is there any progress on the unit tests and e2e tests for https://github.com/kubernetes/kubernetes/pull/31033? Is any help needed?

**Comment by davidopp** [comment]
If you're talking about what's described in #33454 -- it would be great if you are volunteering to do it. If you are referring to something else, let me know.

**Comment by balajismaniam** [comment]
@davidopp  Thanks for pointing me to the issue. I'm willing to work on this but it seems like you have decided not to cherry-pick the tests. (Let me know if that changes and I can work on this.)

**Comment by davidopp** [comment]
Yeah, never mind, we don't need to cherrypick. Sorry for the confusion.

**Comment by erictune** [comment]
Please consider https://github.com/kubernetes/kubernetes/issues/34776 before beta, as it might be a breaking API change to fix it.

**Comment by davidopp** [comment]
ref/ #35318 
ref/ #35324

**Comment by davidopp** [comment]
ref/ kubernetes/kubernetes#25321

**Comment by mikebryant** [comment]
ref #45398

**Comment by rdsubhas** [comment]
As said in #45398, can anyone pls describe/document how to use `PodDisruptionBudget` with helm? Since updates are forbidden, we are forced to manually re-create it all the time which kills source management for this one resource. Does it really count as beta if it doesn't work like any other kubernetes resource with helm?

**Comment by rdsubhas** [comment]
*Copy-paste comment from linked issue*

Hi @davidopp @kow3ns thanks for your responses. Do you know if there is a common standard or style or agreement across kubernetes core resource teams on mutation? Because this is the first example where a resource does not work with helm (and consequently, makes state management and continuous delivery and git-driven changes hard, and requires manual intervention for updating one single resource). It would help a lot if kubernetes resources are helm-friendly, but that means a HUGE techno-social barrier of aligning across resources on the same general standard.

Because certain technical decisions (like immutability) work well when inside a service boundary, but when it leaks outside to the customers/clients/downstreams - it changes client workflows. In this case, it changes helm and CD workflow. Any hopes of getting a mutation policy that is helm-friendly please, for the sake of end-customer workflows, and any hopes of having certain standard expectations when dealing with kubernetes resources and change management / continuous delivery?

**Comment by davidopp** [comment]
Why do you need to update the PDB when using Helm?

**Comment by rdsubhas** [comment]
@davidopp I can give a few examples: Some start with a 3-node ElasticSearch and a certain PDB, then change replicas, and increase the PDB. i.e. Every time someone changes the replicas of a deployment or statefulset, the corresponding PDB numbers are directly or indirectly linked.

**Comment by kow3ns** [comment]
@rdsubhas 

For stateful applications, in the general case, you should not have to modify the maxUnavailable field of the PDB.

1. You should set the maxUnavailable field of a PDB to the maximum number of **planned** disruptions that you are willing to take. For most production deployments of stateful applications this is always 1.
    1. Lets use ZooKeeper as an example. If you deploy an ensemble of 3 servers and plan to tolerate 1 disruption, the implication is that you are okay with loosing quorum when an unplanned disruption occurs concurrently with a planned disruption. Most production deployments will use 5 or more servers. This allows for quorum to be maintained during a single planned disruption, like node maintenance, even when a concurrent unplanned disruption, like node failure, occurs.

1. When you scale a stateful application, the general reasons are to increase capacity or to increase fault tolerance. Here, fault tolerance refers specifically to unplanned disruptions (node, network, power domain failures) that may occur concurrently with planned disruptions. You don't scale an application from 3 to 5 replicas so that you can plan to concurrently kill two of them. You scale it from 3 to 5 to either provide for additional IOPs (read, write, or mixed depending on the application) or to ensure that you can still meet your read and write quorum(s) in the presence of unplanned disruptions.

1. If you really want scale in-variance for your disruption budget, which is a more common use case for serving workloads, you should set your maxUnavaiable to a fixed percentage of the applications capacity (e.g. 33%).

**Comment by davidopp** [comment]
FWIW this discussion has now forked into two parallel threads, here and #45398. I'd suggest coalescing them into one place.

**Comment by rdsubhas** [comment]
@kow3ns 

> most production deployments of stateful applications this is always 1
> If you deploy an ensemble of 3 servers and plan to tolerate 1 disruption
> Most production deployments will use 5 or more servers
> you should set your maxUnavaiable to a fixed percentage of the applications capacity (e.g. 33%)

You have just reinforced change management. Someone starts with 3 servers, sets maxUnavailable to 1 as you said, then goes up to 5 or more servers, sets it to 2, then at some point gets fed up of deleting and changing, sets it to 33%, then scale up more, then increase the quorum, set it to 20%, and so on. Even with the limited example you provided, change management is already clearly the case. And this is just scratching the surface - there is no way that 1 value will work for all apps. Everything else I said - helm, k8s, etc hold stronger than before.

@enisoc has already voiced the broader picture in #45398 hence I am happy to continue the discussion there with @kow3ns.

**Comment by rowleyaj** [comment]
I've commented on #45398 about the helm issue since it was decided discussion would go there.

I wanted to also comment here as well and say this should probably now have the `sig/apps` label and be updated with what is still remaining? I didn't remove sig-scheduling as I'm not sure how much of this was moved to `sig-apps`

Hopefully this works. It should according to https://github.com/kubernetes/test-infra/blob/master/commands.md

/sig apps
